src/eko/harmonics/__init__.py

```diff
@@ -217,3 +218,50 @@ def compute_cache(n, max_weight, is_singlet):
         sx[3, 1:-1] = s4x(n, sx[:, 0], sx[:, -1], is_singlet)
     # return list of list keeping the non zero values
     return [[el for el in sx_list if el != 0] for sx_list in sx]
+
+
+@nb.njit(cache=True)
+def compute_qed_ns_cache(n, s1):
+    r"""Get the harmonics sums cache needed for the qed non singlet AD.
+
+    Parameters
+    ----------
+    n : complex
+        Mellin moment
+    s1 : float
+        harmonic sum :math:`S_1(N)`
+
+    Returns
+    -------
+    list
+        harmonic sums cache. It contains:
+
+        .. math ::
+            [S_1(n/2),
+            S_2(n/2),
+            S_{3}(n/2),
+            S_1((n+1)/2),
+            S_2((n+1)/2),
+            S_{3}((n+1)/2),
+            g_3(n),
+            g_3(n+2)]
```

Comment:
I know this might be painful, so not sure if it's worth it: it should be possible to simplify this in as `g3(n) + something`. 
This is because `g3`  is `S[-2,1,n]` under the hood.

Reply:
Maybe yes, but I tried and I don't think it will be super easy. Anyway I don't think we would get a big improvement in speed being g3 an interpolated function.

Reply:
where are you actually using this? and when you generate the expression for the anomalous dimensions, I guess you have Mathematica in the game any ways, right? so there you can do the replacement easily ... since QED is slow already, we should be careful about calling harmonics - they are the most expensive part (and all of them are in some sense interpolated)

Reply:
I'm using it in the O(as1aem1) and O(aem2) anomalous dimensions and I'm generating it in anomalous_dimensions/__init__. Unfortunately ```S_i(n/2)``` and ```S_i(n/2+1/2)``` cannot be written as a function of ```S_i(n)``` so we need to compute all of them.

Reply:
From what I understand it should be possible to map them into `S_i` and `S_{-i}`, using relations reported [here](https://eko.readthedocs.io/en/latest/theory/Mellin.html#harmonic-sums), but still you will be left with 2 dof.

Reply:
Okay, so
- you use them here
  https://github.com/NNPDF/eko/blob/88e7284b74e3899048b5eb6dfb50f16b583ed512/src/eko/anomalous_dimensions/as1aem1.py#L244
  https://github.com/NNPDF/eko/blob/88e7284b74e3899048b5eb6dfb50f16b583ed512/src/eko/anomalous_dimensions/as1aem1.py#L311
and no where else, correct?
- yes, $S_1(N)$ and $S_1((N+1)/2)$ are linearly independent
- still, from my intuition I would say, that it should be possible to recast all arguments to the list given in https://github.com/NNPDF/eko/blob/88e7284b74e3899048b5eb6dfb50f16b583ed512/src/eko/anomalous_dimensions/as2.py#L16
  because the underlying diagrams/topologies are the same

Reply:
> Okay, so
> 
> * you use them here
>   https://github.com/NNPDF/eko/blob/88e7284b74e3899048b5eb6dfb50f16b583ed512/src/eko/anomalous_dimensions/as1aem1.py#L244
>   https://github.com/NNPDF/eko/blob/88e7284b74e3899048b5eb6dfb50f16b583ed512/src/eko/anomalous_dimensions/as1aem1.py#L311
>   and no where else, correct?

They are also used in : 
https://github.com/NNPDF/eko/blob/feature/qed-solution/src/eko/anomalous_dimensions/aem2.py#L154
https://github.com/NNPDF/eko/blob/feature/qed-solution/src/eko/anomalous_dimensions/aem2.py#L193
and all the others non singlet at O(aem2)

> * still, from my intuition I would say, that it should be possible to recast all arguments to the list given in https://github.com/NNPDF/eko/blob/88e7284b74e3899048b5eb6dfb50f16b583ed512/src/eko/anomalous_dimensions/as2.py#L16
>   because the underlying diagrams/topologies are the same

Yes the same harmonic sums are used at O(as2) and I wanted to pass them also to these anomalous dimensions. But maybe it can be done in another PR since this one is already big enough.

Reply:
> But maybe it can be done in another PR since this one is already big enough.

Agreed

Reply:
Maybe keep a note somewhere